### PR TITLE
Adaboost

### PR DESCRIPTION
**Pull request by parismita**

### CODE REVIEW

**Comment by aldro61** [comment]
I'm not sure why, but there are some problems with the c++ solver on Mac. It works perfectly on linux. I'll take a look.

**Comment by aldro61** [comment]
It's fixed! I'm having a hard time pushing the changes, but I'll do it this weekend.

**Comment by parismita** [comment]
@aldro61 the mmitboost and mmitboost.predict has most of the code same...as we cant find the prediction value without running it, and as its iterative process hence mmitboost cant be used for prediction

So is there any way to improve this?

File: Rpackage/R/mmitboost.R [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    if(all(scores == 0)){
+      break
+    }
+    error <- sum(weights*scores)/sum(weights)
+    alpha <- 0.5 * log( (1. - error) /error)
+    weights <- weights*exp(alpha*scores)
+    final_scores <- final_scores + alpha*prediction
+  }
+  #plot(tree)
```
Comment (aldro61) [inline Rpackage/R/mmitboost.R]: Please remove commented lines.

**Comment by aldro61** [comment]
There might be some code that can be reused, maybe packaged into one or many utility functions. But before that, we should make sure that the implementation is ok.

I think that the Adaboost algorithm that you implemented is for classification. The algorithm is different for regression. Scikit-learn implements the following version (Adaboost.R2) https://www.researchgate.net/publication/2424244_Improving_Regressors_Using_Boosting_Techniques.

Note that according to this algorithm, the training data needs to be resampled with replacement before training each tree. Also, there seems to be a learning rate hyperparameter that can be tuned by the user.

**Comment by parismita** [comment]
ok....i'll check that and update

File: Rpackage/R/mmitboost.R [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  return(result)
+}
+
+.compute_loss <- function(target.mat, prediction, margin, loss){
+  lower = target.mat[,1] - prediction + margin
+  lower[lower < 0] <- 0
+  upper = prediction - target.mat[,2] + margin
+  upper[upper < 0] <- 0
+  if(loss == 'hinge'){
+    cost = lower + upper
```
Comment (parismita) [inline Rpackage/R/mmitboost.R:81]: @aldro61 How to make sure that cost is between 0 to 1?

File: Rpackage/R/mmitboost.predict.R [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#' 
+#' colnames(feature.mat) <- c("a", "b", "c")
+#' feature.mat <- data.frame(feature.mat)
+#' 
+#' trees <- mmitboost(target.mat, feature.mat)
+#' pred <- mmitboost.predict(trees, feature.mat)
+#' 
+#' @export
+mmitboost.predict <- structure(function(target.mat, mmitboost_results) {
+  final_scores <- weighted.mean(mmitboost_results$pred, mmitboost_results$B)
```
Comment (parismita) [inline Rpackage/R/mmitboost.predict.R:31]: @aldro61  How to implement weighted median?

Selection_083
